README.md
# COVID-19 Calculator

[![npm version](https://badge.fury.io/js/%40deptofdefense%2Fcovid19-calculator.svg)](https://badge.fury.io/js/%40deptofdefense%2Fcovid19-calculator)

## Description

COVID-19 Calculator provides a set of scoring algorithms for evaluating COVID-19 symptoms implemented in [TypeScript](https://www.typescriptlang.org/). The algorithms and associated inputs in this model are based on United States Department of Defense medical opinion and are actively used in [MySymptoms.mil](https://mysymptoms.mil/).

This calculator is currently maintained by [Defense Digital Service](https://dds.mil/) (DDS).

## Usage

### Installation

NPM:

```
npm install @deptofdefense/covid19-calculator
```

Yarn:

```
yarn add @deptofdefense/covid19-calculator
```

### Latest Calculator

Importing from the base will import the latest version.

ES6:

```TSX
import { ages, conditions, symptoms, /*...*/ calculate } from '@deptofdefense/covid19-calculator'
```

CommonJS:

```TSX
const { ages, conditions, symptoms, /*...*/ calculate } = require('@deptofdefense/covid19-calculator')
```

### Previous Calculator

If you are interested in importing a previous version simply append the version. Each version functions as an independent package and thus the exports and [TypeScript](https://www.typescriptlang.org/) types may be different between versions.

ES6:

```TSX
import { v1_3_0 } from '@deptofdefense/covid19-calculator'
const { ages, conditions, symptoms, /*...*/ calculate } = v1_3_0
```

CommonJS:

```TSX
const { v1_3_0 } = require('@deptofdefense/covid19-calculator')
const { ages, conditions, symptoms, /*...*/ calculate } = v1_3_0

```

### API

The following functions are exported. A category is a string in the exported list `categories`, i.e., `likelihood`, `exposure`, or `preExisting`. A score is an integer `[0, N)` with the maximum number changing depending on the model version. A level is a human-readable string the indicates the magnitude of the score, e.g., `low`, `medium`, or `high`.

NOTE: At this time, consider this api in "beta" with the possibility of changing slightly in future releases.

```TSX
import { calculate, calculateScores, calculateLevel, calculateLevels } from '@deptofdefense/covid19-calculator'
// calculate returns the integer scores and string levels
const { scores, levels } = calculate({ ... })
// calculateScores just returns the integer scores
const { likelihoodScore, exposureScore, preExistingScore } = calculateScores({
  symptoms: [...],
  conditions: [...]
})
// calculateLevels returns the string levels for the given scores
const { likelihoodLevel, exposureLevel, preExistingLevel } = calculateLevels({
  likelihoodScore,
  exposureScore,
  preExistingScore
})
// calculateLevel returns the string level for the given score
const { likelihoodLevel } = calculateLevel('likelihood', 30)
```

If you need to correlate incoming data with calculator parameters, consider using a hashing algorithm to resolve parameters. See the [advanced example](examples/advanced.js) for more info.

## Examples

For the full list of examples see the [examples](examples) folder.

### Simple Example

```TSX
const { calculate } = require('@deptofdefense/covid19-calculator');

const result = calculate({
  symptoms: ['fever', 'lossOfSmell'],
  conditions: ['hardBreathing'],
});
/*
 * result
 * {
 *   scores: { likelihood: 20, exposure: 0, preExisting: 0 },
 *   levels: { likelihood: 'low', exposure: 'low', preExisting: 'low' }
 * }
 */
```

## Current Model

### Data

#### Ages

- What's your age (in years)?
  - 0-17 - preExisting: 0
  - 18-39 - preExisting: 0
  - 40-64 - preExisting: 0
  - 65-69 - preExisting: 10
  - 70-79 - preExisting: 15
  - 80-Infinity - preExisting: 20

#### Exposures

- In the last 14 days, did you have close contact with someone with symptoms of COVID-19 or diagnosed with COVID-19?
  - contact - Yes - exposure: 30
  - possibleContact - I Don't Know - exposure: 10
  - noContact - No - exposure: 0

#### Symptoms

- Select your current symptoms:
  - cough - Cough - likelihood: 20
  - sob - Shortness of breath - likelihood: 20
  - fever - Fever - likelihood: 20
  - bodyAches - Body aches or muscle pain - likelihood: 10
  - headache - Headache - likelihood: 10
  - throat - Sore throat - likelihood: 10
  - chills - Chills or repeated shaking with chills - likelihood: 10
  - lossOfSmell - New loss of taste or smell - likelihood: 20
  - vomiting - Nausea or vomiting (aliases: Vomiting) - likelihood: 5
  - dizziness - Dizziness or lightheadedness - likelihood: 5
  - diarrhea - Diarrhea - likelihood: 5
  - fatigue - Fatigue - likelihood: 5
  - congestion - Congestion or runny nose - likelihood: 5

#### Conditions

- Select current conditions:
  - arouse - Disoriented or inability to wake/stay awake (worse than sleepy) (aliases: 'New confusion or inability to arouse', 'Confusion or inability to wake / stay awake') - likelihood: 40
  - hardBreathing - Trouble or difficulty breathing - likelihood: 40
  - chestPain - Persistent pain or pressure in the chest - likelihood: 40
  - blueLips - Bluish lips or face - likelihood: 40

#### Pre-existing Conditions

- Select pre-existing conditions:
  - diabetesType2 - Type 2 diabetes mellitus - preExisting: 20
  - heartDisease - Heart conditions, such as heart failure, coronary artery disease, or cardiomyopathies - preExisting: 20
  - cancer - Cancer - preExisting: 20
  - kidneyDisease - Chronic kidney disease - preExisting: 20
  - copd - COPD - preExisting: 20
  - organTransplant - Had a solid organ transplant - preExisting: 20
  - sickleCellDisease - Sickle cell disease - preExisting: 20
  - overweight - Very overweight (Body Mass Index (BMI) of 30 or higher) - preExisting: 20
  - asthma - Asthma - preExisting: 15
  - smoke - Currently smoking - preExisting: 15
  - steroids - Take corticosteroids or other immunosuppressive medications - preExisting: 15
  - pregnant - Pregnant - preExisting: 15
  - cerebrovascular - Cerebrovascular disease - preExisting: 15
  - hypertension - Hypertension - preExisting: 15
  - diabetesType1 - Type 1 diabetes - preExisting: 10
  - lungDisease - Cystic fibrosis, emphysema, or other chronic lung diseases - preExisting: 10
  - thalassemia - Thalassemia - preExisting: 10
  - neurologicCondition - Neurologic conditions - preExisting: 10
  - liverDisease - Liver disease - preExisting: 10
  - compromisedImmuneSystem - Compromised immune system - preExisting: 10
  - hiv - HIV - preExisting: 10
  - boneMarrowTransplant - Had a bone marrow transplant - preExisting: 10

### Scales

#### Likelihood

- veryLow: 0 - 9
- low: 10 - 19
- medium: 20 - 39
- high: 40 - 69
- veryHigh: 70+

#### PreExisting

- low: 0 - 9
- medium: 10 - 29
- high: 30+

#### Exposure

- low: 0 - 9
- medium: 10 - 29
- high: 30+

## Local Development

This project was bootstrapped with [TSDX](https://github.com/jaredpalmer/tsdx). Below is a list of commands you will probably find useful.

### `npm start` or `yarn start`

Runs the project in development/watch mode. Your project will be rebuilt upon changes. TSDX has a special logger for you convenience. Error messages are pretty printed and formatted for compatibility VS Code's Problems tab. Your library will be rebuilt if you make edits.

### `npm run build` or `yarn build`

Bundles the package to the `dist` folder. The package is optimized and bundled with Rollup into multiple formats (CommonJS, UMD, and ES Module).

### `npm test` or `yarn test`

Runs the test watcher (Jest) in an interactive mode. By default, runs tests related to files changed since the last commit.

## Contributing

We'd love to have your contributions! Please see [CONTRIBUTING.md](CONTRIBUTING.md) for more info.

## Security

Please see [SECURITY.md](SECURITY.md) for more info.

## License

This project constitutes a work of the United States Government and is not subject to domestic copyright protection under 17 USC § 105. However, because the project utilizes code licensed from contributors and other third parties, it therefore is licensed under the MIT License. See LICENSE file for more information.


CHANGELOG.md
# Change Log

<!--
Please use the following format for version changes
## [VERSION] - DATE

### Added

### Changed

### Removed

### Fixed

For more details see the specific version's [README](https://github.com/deptofdefense/covid19-calculator/blob/master/src/{{VERSION}}/README.md)

-->

## 1.4.0 - 4/27/2021

Small change to the description blueLips

### Changes

- blueLips - 'Bluish lips or face' +10 -> 'Pale, gray, or blue-colored skin, lips, or nail beds '

For more details see the specific version's [README](https://github.com/deptofdefense/covid19-calculator/blob/master/src/v1.10.1/README.md)

## 1.3.0 - 4/13/2021

Small change to the description and pointing of lossOfSmell

### Changes

- lossOfSmell - 'Altered smell/taste' +10 -> 'New loss of taste or smell' +20

For more details see the specific version's [README](https://github.com/deptofdefense/covid19-calculator/blob/master/src/v1.10.0/README.md)

## 1.2.0 - 11/5/2020

Small change to the description of arouse

### Changes

- arouse - 'Confusion or inability to wake / stay awake' -> 'Disoriented or inability to wake/stay awake (worse than sleepy)'

For more details see the specific version's [README](https://github.com/deptofdefense/covid19-calculator/blob/master/src/v1.9.1/README.md)

## 1.1.0 - 9/22/2020

Adjustments have been made in accordnace with recent studies.

### Changes

Symptoms:

- (Changed) fever - Fever +10 -> +20

Pre-Existing Conditions:

Across the board pre-exisitng conditions have been split apart. These are all marked as change because they all map back to a previous pre-existing condition.

- (Changed) diabetesType2 - Type 2 diabetes mellitus - preExisting: 20
- (Changed) heartDisease - Heart conditions, such as heart failure, coronary artery disease, or cardiomyopathies - preExisting: 20
- (Changed) cancer - Cancer - preExisting: 20
- (Changed) kidneyDisease - Chronic kidney disease - preExisting: 20
- (Changed) copd - COPD - preExisting: 20
- (Changed) organTransplant - Had a solid organ transplant - preExisting: 20
- (Changed) sickleCellDisease - Sickle cell disease - preExisting: 20
- (Changed) overweight - Very overweight (Body Mass Index (BMI) of 30 or higher) - preExisting: 20
- (Changed) asthma - Asthma - preExisting: 15
- (Changed) smoke - Currently smoking - preExisting: 15
- (Changed) steroids - Take corticosteroids or other immunosuppressive medications - preExisting: 15
- (Changed) pregnant - Pregnant - preExisting: 15
- (Changed) cerebrovascular - Cerebrovascular disease - preExisting: 15
- (Changed) hypertension - Hypertension - preExisting: 15
- (Changed) diabetesType1 - Type 1 diabetes - preExisting: 10
- (Changed) lungDisease - Cystic fibrosis, emphysema, or other chronic lung diseases - preExisting: 10
- (Changed) thalassemia - Thalassemia - preExisting: 10
- (Changed) neurologicCondition - Neurologic conditions - preExisting: 10
- (Changed) liverDisease - Liver disease - preExisting: 10
- (Changed) compromisedImmuneSystem - Compromised immune system - preExisting: 10
- (Changed) hiv - HIV - preExisting: 10
- (Changed) boneMarrowTransplant - Had a bone marrow transplant - preExisting: 10

For more details see the specific version's [README](https://github.com/deptofdefense/covid19-calculator/blob/master/src/v1.9.0/README.md)

## 1.0.0 - 9/15/2020

### BREAKING CHANGES

All options now have a slightly different shape and include a label. This label is intented to provide more information on how the options are intended to be used and also solves the problem of trying to have a one of option. To access the same information, use the `items` key inside of each `ages`, `symptoms`, `conditions`, etc.

#### Previously

```
export const ages: ScorableCollection<Age> = [
  { prompt: '0-17', key: '0-17', preExisting: 0 },
  { prompt: '18-39', key: '18-39', preExisting: 0 },
  { prompt: '40-64', key: '40-64', preExisting: 0 },
  { prompt: '65-69', key: '65-69', preExisting: 10 },
  { prompt: '70-79', key: '70-79', preExisting: 15 },
  { prompt: '80+', key: '80-Infinity', preExisting: 20 },
];
```

#### Now

```
export const ages: ScorableCollection<Age> = {
  key: 'ages',
  label: "What's your age (in years)?",
  items: [
    { prompt: '0-17', key: '0-17', preExisting: 0 },
    { prompt: '18-39', key: '18-39', preExisting: 0 },
    { prompt: '40-64', key: '40-64', preExisting: 0 },
    { prompt: '65-69', key: '65-69', preExisting: 10 },
    { prompt: '70-79', key: '70-79', preExisting: 15 },
    { prompt: '80+', key: '80-Infinity', preExisting: 20 },
  ],
};
```

### Changes

Calulcator version 1.8.0 includes wording changes to exopsure questions and 1.7.0 wording was added as in aliases.

- In the last two weeks, did you care for or have close contact (within 6 feet of an infected person for at least 15 minutes) with someone with symptoms of COVID-19, tested for COVID-19, or diagnosed with COVID-19?
  - contact - Yes - exposure: 30
  - possibleContact - I Don't Know - exposure: 10
  - noContact - No - exposure: 0

For more details see the specific version's [README](https://github.com/deptofdefense/covid19-calculator/blob/master/src/v1.8.0/README.md)

## 0.6.0 - 8/31/2020

### Changes

Calulcator version 1.7.0 includes updates to exopsure questions.

- (Added) New questions asking more directly about exposure to someone with COVID-19.
- (Removed) The three previous travel questions

For more details see the specific version's [README](https://github.com/deptofdefense/covid19-calculator/blob/master/src/v1.7.0/README.md)

## 0.5.1 - 7/30/2020

### Fixed

Calculator version 1.6.0 was previously reporting a `calculatorVersion` of `'1.5.0'`. This has been resolved.

## 0.5.0 - 7/8/2020

### Changes

Calculator version 1.6.0 includes updates based on recent CDC updated recommendations. 1.6.0 includes the following:

- (Added) Congestion or runny nose +5
- (Changed) Nausea or vomiting +5 (previously was just "Vomiting")
- (Removed) Itchy, red or pink eyes +5

## 0.4.0 - 7/2/2020

### Changes


src/v1.3.0/scale.ts
import {
  ScoreCategory,
  ScorableCollection,
  ScoredData,
  LevelData,
} from './types';
import { ages } from './ages';
import { conditions } from './conditions';
import { exposures } from './exposures';
import { preExistingConditions } from './preExistingConditions';
import { symptoms } from './symptoms';

const scoreData = {
  ages,
  conditions,
  exposures,
  preExistingConditions,
  symptoms,
};

export const categories: ScoreCategory[] = [
  'likelihood',
  'exposure',
  'preExisting',
];
export const emptyScore: ScoredData = {
  likelihood: 0,
  preExisting: 0,
  exposure: 0,
};

type ScaleLevels = Readonly<Record<ScoreCategory, [string, number][]>>;

//TODO: It would be to have the types fall from the `scale` data
export type LikelihoodLevel =
  | 'veryLow'
  | 'low'
  | 'medium'
  | 'high'
  | 'veryHigh';
export type PreExitingLevel = 'low' | 'medium' | 'high';
export type ExposureLevel = 'low' | 'medium' | 'high';

export const scale: ScaleLevels = {
  likelihood: [
    ['veryLow', 0],
    ['low', 10],
    ['medium', 20],
    ['high', 40],
    ['veryHigh', 70],
  ],
  preExisting: [
    ['low', 0],
    ['medium', 10],
    ['high', 30],
  ],
  exposure: [
    ['low', 0],
    ['medium', 10],
    ['high', 30],
  ],
};

type DataType = typeof scoreData;
export type DataToScore = Partial<
  {
    [K in keyof DataType]: string[];
  }
>;

const mergeScores = (...args: Partial<ScoredData>[]): ScoredData =>
  args.reduce<ScoredData>(
    (total, next) => ({
      likelihood: (total.likelihood ?? 0) + (next.likelihood ?? 0),
      exposure: (total.exposure ?? 0) + (next.exposure ?? 0),
      preExisting: (total.preExisting ?? 0) + (next.preExisting ?? 0),
    }),
    emptyScore
  );

export const calculateScore = (
  ScoreCategory: keyof typeof scoreData,
  values: string[]
): ScoredData => {
  const collection = scoreData[ScoreCategory] as ScorableCollection<string>;
  return mergeScores(...collection.filter(item => values.includes(item.key)));
};

export const calculateScores = (dataToScore: DataToScore) => {
  const {
    ages = [],
    exposures = [],
    preExistingConditions = [],
    conditions = [],
    symptoms = [],
  } = dataToScore;

  const scores = mergeScores(
    calculateScore('ages', ages),
    calculateScore('exposures', exposures),
    calculateScore('preExistingConditions', preExistingConditions),
    calculateScore('conditions', conditions),
    calculateScore('symptoms', symptoms)
  );

  return scores;
};

export const calculateLevel = (category: ScoreCategory, score: number) => {
  if (scale[category] === undefined) {
    throw new Error(
      `Category is unknown. Must be one of: ${categories.join(', ')}`
    );
  }
  const [lowest, ...rest] = scale[category];
  let result = lowest[0];

  for (let [level, threshold] of rest) {
    if (score >= threshold) {
      result = level;
    }
  }
  // return lowest level
  return result;
};

export const calculateLevels = (scores: ScoredData): LevelData =>
  categories.reduce((levels, category) => {
    return {
      ...levels,
      [category]: calculateLevel(category, scores[category] ?? 0),
    };
  }, {} as LevelData);

export const calculate = (dataToScore: DataToScore) => {
  const scores = calculateScores(dataToScore);

  const levels = calculateLevels(scores);

  return { scores, levels };
};
